Mult decrease bound

goelvidhi opened the pull request:
Hope this should remove any ambiguity we had in how cwnd_start is set during multiplicative decrease.

Also, updated some minor things in change log.

larseggert commented on draft-eggert-tcpm-rfc8312bis.md:
```diff
@@ -619,10 +616,12 @@ for each received ACK, where *target* is calculated as described in
-reduces *cwnd* and *ssthresh* immediately as described below. For both
-packet loss and congestion detection through ECN, the sender MAY
+reduces *cwnd* and *ssthresh* immediately as described below. An
+implementation MAY set a smaller ssthresh than suggested below to
```
```suggestion
implementation MAY set a smaller *ssthresh* than suggested below to
```

larseggert commented on draft-eggert-tcpm-rfc8312bis.md:
```diff
@@ -619,10 +616,12 @@ for each received ACK, where *target* is calculated as described in
+For both packet loss and congestion detection through ECN, the sender MAY
 employ a Fast Recovery algorithm to gradually adjust the congestion
-window to its new reduced value. The parameter {{{β}{}}}*<sub>cubic</sub>*
+window to its new reduced ssthresh value. The parameter {{{β}{}}}*<sub>cubic</sub>*
```
```suggestion
window to its new reduced *ssthresh* value. The parameter {{{β}{}}}*<sub>cubic</sub>*
```

larseggert commented on draft-eggert-tcpm-rfc8312bis.md:
```diff
@@ -989,22 +988,25 @@ Richard Scheffenegger and Alexander Zimmermann originally co-authored
 - fix a mistake in *W<sub>max</sub>* calculation in the fast convergence section.
   ([#51](https://github.com/NTAP/rfc8312bis/issues/51))
   
+- clarity on setting ssthresh and *cwnd<sub>start</sub>* during multiplicative decrease.
```
```suggestion
- clarity on setting *ssthresh* and *cwnd<sub>start</sub>* during multiplicative decrease.
```

goelvidhi commented:
@larseggert not sure why the last commit failed the checks. The change was ssthresh to *ssthresh* in the change log

larseggert commented:
`draft-eggert-tcpm-rfc8312bis.md:996:81 MD013/line-length Line length [Expected: 80; Actual: 90]`

lisongxu commented:
How about changing "*cwnd<sub>start</sub>* is calculated as described in {{mult-dec}} when a congestion event is detected." to ""*cwnd<sub>start</sub>* is the new *cwnd* calculated when a congestion event is detected as described in {{mult-dec}}."?

lisongxu reviewed (changes requested):
How about changing "cwndstart is calculated as described in {{mult-dec}} when a congestion event is detected." to ""cwndstart is the new cwnd calculated when a congestion event is detected as described in {{mult-dec}}."?

goelvidhi commented:
> How about changing "cwndstart is calculated as described in {{mult-dec}} when a congestion event is detected." to ""cwndstart is the new cwnd calculated when a congestion event is detected as described in {{mult-dec}}."?

I updated the text.
Also, I removed the below redundant line. We already describe cwnd_start in variable definition.

> where *cwnd<sub>start</sub>* is the congestion window at the beginning
 of the current congestion avoidance stage.

 I also updated the Multiplicative Decrease section to call the reduced window as cwnd_start. 

 Let me know if these updates are ok.

lisongxu reviewed (changes requested):
line 472: change "where *cwnd<sub>start</sub>* is the new congestion window calculated when a congestion event is detected as described in {{mult-dec}}" to "where *cwnd<sub>start</sub>* is the new *cwnd* calculated when the previous congestion event was detected as described in {{mult-dec}}" ?

line 631: "cwnd_{start} = ssthresh " should be "cwnd = ssthresh ". 

Thanks

goelvidhi commented:
> line 472: change "where _cwndstart_ is the new congestion window calculated when a congestion event is detected as described in {{mult-dec}}" to "where _cwndstart_ is the new _cwnd_ calculated when the previous congestion event was detected as described in {{mult-dec}}" ?

Previous could mean previous to the last congestion event. Should we say ..when the last congestion event was detected ...?
> 
> line 631: "cwnd_{start} = ssthresh " should be "cwnd = ssthresh ".
> 
> Thanks

I changed it to cwnd_{start} for clarity. Could you tell why do you think "cwnd = ssthresh " is better?

lisongxu commented:
Thanks, Vidhi. This is because cwnd is the one used everywhere. If line 631 is "cwnd_{start} = ssthresh ", then variable cwnd is not defined.

lisongxu commented:
for cwnd_start at line 472, I still like the definition that we discussed before. that is  "where cwnd_start is the cwnd at the very beginning of the current congestion avoidance stage".  This covers all possible cases. 
   * loss event, congestion avoidance
   * TCP starts, slow start, congestion avoidance
   * timeout, slow start, congestion avoidance

How do you think? Thanks

goelvidhi commented:
ok. I will revert the change to like 631.

For line 472, how about the below text?
where *cwnd<sub>start</sub>* is the congestion window at the beginning
of the current congestion avoidance stage. When a congestion event is
detected, *cwnd<sub>start</sub>* is calculated as described in {{mult-dec}}.

lisongxu commented:
I added "for example" so that a reader knows this is only one of many possible cases.

where cwndstart is the congestion window at the beginning of the current congestion avoidance stage. For example, right after a congestion event, cwndstart is equal to the new cwnd calculated as described in {{mult-dec}}.

goelvidhi commented:
> I added "for example" so that a reader knows this is only one of many possible cases.
> 
> where cwndstart is the congestion window at the beginning of the current congestion avoidance stage. For example, right after a congestion event, cwndstart is equal to the new cwnd calculated as described in {{mult-dec}}.

sounds good. Updated.

goelvidhi commented:
> for cwnd_start at line 472, I still like the definition that we discussed before. that is "where cwnd_start is the cwnd at the very beginning of the current congestion avoidance stage". This covers all possible cases.
> 
> * loss event, congestion avoidance
> * TCP starts, slow start, congestion avoidance
> * timeout, slow start, congestion avoidance

The other scenarios where we go from slow start to CA without loss have some caveats which are not mentioned in the draft.
For example, K should be 0 and W_max should be set to cwnd. Do you think we should add some text?

lisongxu commented:
> The other scenarios where we go from slow start to CA without loss have some caveats which are not mentioned in the draft.
> For example, K should be 0 and W_max should be set to cwnd. Do you think we should add some text?


That case is discussed in Section 4.10 "slow start"?

"When CUBIC uses hybrid slow start [HR08], it may exit the first slow start without incurring any packet loss and thus Wmax is undefined. In this special case, CUBIC switches to congestion avoidance and increases its congestion window size using Figure 1, where t is the elapsed time since the beginning of the current congestion avoidance, K is set to 0, and Wmax is set to the congestion window size at the beginning of the current congestion avoidance stage."

lisongxu reviewed (approved):
Thank you, Vidhi!

goelvidhi commented:
> > The other scenarios where we go from slow start to CA without loss have some caveats which are not mentioned in the draft.
> > For example, K should be 0 and W_max should be set to cwnd. Do you think we should add some text?
> 
> That case is discussed in Section 4.10 "slow start"?
> 
> "When CUBIC uses hybrid slow start [HR08], it may exit the first slow start without incurring any packet loss and thus Wmax is undefined. In this special case, CUBIC switches to congestion avoidance and increases its congestion window size using Figure 1, where t is the elapsed time since the beginning of the current congestion avoidance, K is set to 0, and Wmax is set to the congestion window size at the beginning of the current congestion avoidance stage."

Ah.. got it. Looks good.

goelvidhi commented:
@larseggert this PR is ready to merge.
